### PR TITLE
Add controller example

### PR DESCRIPTION
**Pull request by thumbnail**
Todo
 - [x]  use `PlayerCommandRing` to average multiple commands on the same player.
 - [x]  update subra
 - [ ] ~allow arguments to define controller bindings.~ separate issue
 - [x] Move `StrategySupplier` to a separate class.
 - [x] simplify mapping from controllerstates to strategy
 - [x] sensical defaults
 - [x] update readme

Arguments can be passed to the application in the following syntax `<property>: <value>`

| Property  | description  | default |
|---|---|---|
| Ip   | multicast ip  | `224.0.0.1` |
| Port  | multicast port  | `10001` |

### CODE REVIEW

**Comment by romni** [comment]
readme title needs to be updated from `SSL Controller Example` to `subra-examples

File: src/main/java/io/leonis/example/ControllerExample.java [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+ */
+public class ControllerExample {
+
+  private final static Map<String, String> DEFAULTS = ImmutableMap.of(
+      "port", "10001",
+      "ip", "224.0.0.1");
+
+  /**
+   * Constructs a new ControllerExample which submits {@link io.leonis.subra.protocol.Robot generated
+   * commands} to multicast on the supplied ip and port.
```
Comment (romni) [inline src/main/java/io/leonis/example/ControllerExample.java]: capitalize IP

File: src/main/java/io/leonis/example/ControllerExample.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+      final int port,
+      final Map<JamepadControllerIdentity, Set<PlayerIdentity>> controllerMapping
+  ) throws IOException {
+    // the controller handler which parses the active controls
+    final Function<JamepadController, PlayerCommand> handler = new JamepadControllerHandler();
+
+    // create a stream of gamepad states based on the mapping
+    Flux.from(new JamepadPublisher(controllerMapping))
+        .map(MappingSupplier::getAgentMapping)
+        // for each controller,
```
Comment (romni) [inline src/main/java/io/leonis/example/ControllerExample.java:54]: i think this comment was going somewhere 😛

File: src/main/java/io/leonis/ipc/CliSettings.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+@AllArgsConstructor
+public final class CliSettings implements Function<String[], Map<String, String>> {
+
+  /**
+   * Default settings. note that all keys should be present
+   */
+  private final Map<String, String> defaults;
+
+  /**
+   * @param args Array of strings in {@code <key>:<value>}-format
```
Comment (romni) [inline src/main/java/io/leonis/ipc/CliSettings.java:26]: missing period

File: src/main/java/io/leonis/ipc/CliSettings.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * A {@link Function} which transforms a {@link String[] string array} to a {@link Map} merged with
+ * the provided defaults. Extra keys will be ignored.
+ *
+ * @author Jeroen de Jong
+ */
+@AllArgsConstructor
+public final class CliSettings implements Function<String[], Map<String, String>> {
+
+  /**
+   * Default settings. note that all keys should be present
```
Comment (romni) [inline src/main/java/io/leonis/ipc/CliSettings.java:21]: `Note` and missing period

File: src/main/java/io/leonis/ipc/CliSettings.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+   * @return The input as a map merged with the existing defaults.
+   */
+  @Override
+  public Map<String, String> apply(final String[] args) {
+    Map<String, String> input = Stream.of(args)
+        .map(arg -> arg.split(":"))
+        .collect(Collectors.toMap(strings -> strings[0], strings -> strings[1]));
+    return this.defaults.entrySet().stream().collect(Collectors.toMap(
+        Entry::getKey,
+        def -> input.getOrDefault(def.getKey(), def.getValue())));
```
Comment (romni) [inline src/main/java/io/leonis/ipc/CliSettings.java:36]: is there maybe a better name for `def`? like `defaults` or `argAndDefault`?

File: build.gradle [hunk 6]
```diff
@@ -4,7 +4,7 @@ apply plugin: 'java'
 
 group 'io.leonis'
 version '0.0.1'
-archivesBaseName = "ssl-controller-example"
+archivesBaseName = "subra-examples"
 rootProject.description = 'Reference implementation for controlling Small Size League robots with an Xbox controller.'
```
Comment (romni) [inline build.gradle]: needs updating to match what's in the readme

File: README.md [hunk 7]
```diff
@@ -8,6 +8,21 @@ You'll need at least Java 1.8 ([jre](https://www.java.com/download/)
...
+
+## Controller Example
+The class `ControllerExample` can be found in the example package. Currently the controller mapping
+is hardcoded in the `main`-method of the example, which binds robot 1 and 2 of the blue team to the
+first controller that is found.
+
+### settings
+| Property  | description  | default |
+|---|---|---|
+| Ip   | multicast ip  | `224.0.0.1` |
```
Comment (romni) [inline README.md]: `IP`

File: README.md [hunk 8]
```diff
@@ -8,6 +8,21 @@ You'll need at least Java 1.8 ([jre](https://www.java.com/download/)
...
+## Configuration
+Most examples accept arguments in the following format: `<property>:<value>`. For example setting
+the ip would be `ip:123.123.123.123`.
+
+## Controller Example
+The class `ControllerExample` can be found in the example package. Currently the controller mapping
+is hardcoded in the `main`-method of the example, which binds robot 1 and 2 of the blue team to the
+first controller that is found.
+
+### settings
```
Comment (romni) [inline README.md]: `Settings`

File: README.md [hunk 9]
```diff
@@ -8,6 +8,21 @@ You'll need at least Java 1.8 ([jre](https://www.java.com/download/)
 /[jdk](http://www.oracle.com/technetwork/java/javase/downloads/index-jsp-138363.html))
 to run `ssl-controller-example`.
 
+## Configuration
+Most examples accept arguments in the following format: `<property>:<value>`. For example setting
+the ip would be `ip:123.123.123.123`.
```
Comment (romni) [inline README.md]: IP

File: README.md [hunk 10]
```diff
@@ -8,6 +8,21 @@ You'll need at least Java 1.8 ([jre](https://www.java.com/download/)
 /[jdk](http://www.oracle.com/technetwork/java/javase/downloads/index-jsp-138363.html))
 to run `ssl-controller-example`.
 
+## Configuration
+Most examples accept arguments in the following format: `<property>:<value>`. For example setting
```
Comment (romni) [inline README.md]: `For example, setting`

File: src/main/java/io/leonis/example/ControllerExample.java (hunk 2 above)
Comment (thumbnail) [reply src/main/java/io/leonis/example/ControllerExample.java:54]: you removed it :')

**Review by romni — APPROVED** [review]

**Pull request merged**